src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -515,46 +502,106 @@ public void ClearValidationState(string key)
             }
         }
 
-        private ModelStateEntry GetModelStateForKey(string key)
+        private ModelStateNode GetNode(string key) => GetNode(key, createIfNotExists: false);
+
+        private ModelStateNode GetOrAddNode(string key) => GetNode(key, createIfNotExists: true);
+
+        private ModelStateNode GetNode(string key, bool createIfNotExists)
         {
-            if (key == null)
+            Debug.Assert(key != null);
+            if (key.Length == 0)
             {
-                throw new ArgumentNullException(nameof(key));
+                return _root;
             }
 
-            ModelStateEntry entry;
-            if (!TryGetValue(key, out entry))
+            // For a key of the format, foo.bar[0].baz["qux"] we'll create the following nodes:
+            // foo
+            //  -> bar
+            //   -> [0]
+            //    -> baz
+            //     -> ["qux"]
```

Comment:
Why does this still use an unsupported key? Key will be `[qux]` in any real scenario -- never quoted.